// deck.gl/dist/test-utils/luma.gl/io-basic/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseModel = exports.loadModel = exports.loadProgram = exports.loadTextures = exports.loadImages = exports.loadFiles = exports.setPathPrefix = undefined;

var _setPathPrefix = require('./set-path-prefix');

Object.defineProperty(exports, 'setPathPrefix', {
  enumerable: true,
  get: function get() {
    return _setPathPrefix.setPathPrefix;
  }
});

var _loadFiles = require('./load-files');

Object.defineProperty(exports, 'loadFiles', {
  enumerable: true,
  get: function get() {
    return _loadFiles.loadFiles;
  }
});
Object.defineProperty(exports, 'loadImages', {
  enumerable: true,
  get: function get() {
    return _loadFiles.loadImages;
  }
});

var _loadTextures = require('./load-textures');

Object.defineProperty(exports, 'loadTextures', {
  enumerable: true,
  get: function get() {
    return _loadTextures.loadTextures;
  }
});
Object.defineProperty(exports, 'loadProgram', {
  enumerable: true,
  get: function get() {
    return _loadTextures.loadProgram;
  }
});

var _jsonLoader = require('./json-loader');

Object.defineProperty(exports, 'loadModel', {
  enumerable: true,
  get: function get() {
    return _jsonLoader.loadModel;
  }
});
Object.defineProperty(exports, 'parseModel', {
  enumerable: true,
  get: function get() {
    return _jsonLoader.parseModel;
  }
});

var _globals = require('../utils/globals');

// Call a require based helper to select platform to export
if (_globals.isBrowser) {
  module.exports.loadFile = require('browser-request-file');
  module.exports.loadImage = require('browser-request-image');
  module.exports.readFile = require('browser-read-file');
} else {
  var fs = module.require('fs');
  // TODO - needs to be promisified...
  module.exports.readFile = fs && fs.readFile;
}
